src/server/db_slice.cc

```diff
@@ -1096,10 +1109,79 @@ auto DbSlice::DeleteExpiredStep(const Context& cntx, unsigned count) -> DeleteEx
   return result;
 }
 
-// TODO: Design a better background evicting heuristic.
+int32_t DbSlice::GetNextSegmentForEviction(int32_t segment_id, DbIndex db_ind) const {
+  // wraps around if we reached the end
+  return db_arr_[db_ind]->prime.NextSeg((size_t)segment_id) %
+         db_arr_[db_ind]->prime.GetSegmentCount();
+}
+
 void DbSlice::FreeMemWithEvictionStep(DbIndex db_ind, size_t increase_goal_bytes) {
-  if (!caching_mode_)
+  DCHECK(!owner_->IsReplica());
+  if ((!caching_mode_) || !expire_allowed_ || !GetFlag(FLAGS_enable_heartbeat_eviction))
     return;
+
+  auto max_eviction_per_hb = GetFlag(FLAGS_max_eviction_per_heartbeat);
+  auto max_segment_to_consider = GetFlag(FLAGS_max_segment_to_consider);
+
+  auto time_start = absl::GetCurrentTimeNanos();
+  auto& db_table = db_arr_[db_ind];
+  int32_t num_segments = db_table->prime.GetSegmentCount();
+  int32_t num_buckets = PrimeTable::Segment_t::kTotalBuckets;
+  int32_t num_slots = PrimeTable::Segment_t::kNumSlots;
+
+  size_t used_memory_after;
+  size_t evicted = 0;
+  string tmp;
+  int32_t starting_segment_id = rand() % num_segments;
+  size_t used_memory_before = owner_->UsedMemory();
+  for (int32_t slot_id = num_slots - 1; slot_id >= 0; --slot_id) {
```

Comment:
can you please add FiberAtomicGuard before this for loop to make sure we dont have any preemption in one of the functions calls

Reply:
added. 

I didn't know preemption could happen, so could you explain when will preemption happen, or explain an example? i'm just curious.

Reply:
Fiber preemption can happen because we now write into a journal queue so just adding the guard is not enough because we CAN preempt.  @theyueli  Adi is correct in identifying the following use case: you iterate a data structure and you possibly perform I/O and deletions within the loop. a Fiber can preempt and possibly another fiber can mutate this data structure. As a result you will be iterating on possibly inconsistent data. So to avoid this, you need to separate logically the code into two parts: the first one iterates and deletes the items and gathers their keys into a vector. And then the second part that sends the deletion and by the way we can send it as a single command with all the keys. this way we separate the code into the atomic section and the journaling (/O) . the first part should have the atomic guard.

Reply:
@romange sounds good, and thanks for the explanation! which function shall I use to journal a vector of strings in one shot?

Reply:
@romange something like

```
journal->RecordEntry(0, journal::Op::EXPIRED, dbid, 1, make_pair("DEL", ArgSlice{keys}), false);
```
where I append all the keys into `keys` ?

Reply:
I dont think this logic is needed @romange as we should not preept in the flow of RecordExpiry, in this function we pass a flag there to not await

Reply:
I see the flag now though I do not understand how it can be - what if we do not have enough space in the channel?
Sounds like something we will have to fix anyway to limit memory usage in case the the replica socket is slow or stuck

Reply:
In any case, I think it's a good thing to do because I believe we will need to introduce backpressure in the near future.

Reply:
Looking at the code further, seems that `BufferedStreamerBase::producer_buf_` capacity should also be tracked as part of our effort to attribute major memory spenders @adiholden

Reply:
@romange @adiholden  Adi and I just chatted on this. I just made the change to separate them into two stages. @romange please take another look on the separation.